plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -357,15 +357,48 @@ maybeAddSubst body a s n rhs = do
     hints <- view iiHints
     let hinted = shouldInline hints a n
 
-    preUnconditional <- preInlineUnconditional rhs'
-    if preUnconditional
+    if hinted -- if we've been told specifically, then do it right away
     then extendAndDrop (Done $ dupable rhs')
     else do
-        -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-        postUnconditional <- postInlineUnconditional rhs'
-        if hinted || postUnconditional
+        -- See Note [Inlining and purity]
+        termIsPure <- checkPurity rhs'
+        let preInlineUnconditional :: InlineM tyname name uni fun a Bool
+            preInlineUnconditional = do
+                usgs <- view iiUsages
+                -- 'inlining' terms used 0 times is a cheap way to remove dead code while we're here
+                let termUsedAtMostOnce = Usages.getUsageCount n usgs <= 1
+                if not termUsedAtMostOnce -- if the term is used more than once, we don't inline
+                then pure False
+                else do -- otherwise, check if it's nonstrict first
+                            -- effectSafe = case s of
+                                -- Strict    -> termIsPure || immediatelyEvaluated
+                                -- NonStrict -> True
+                    if s == NonStrict
+                    then pure True -- if it's nonstrict, we inline
```

Comment:
I don't believe this is necessary. There is no `IO` in `InlineM`, and using `||` and `&&` should be able to short circuit and avoid the expensive computation. The current implementation does call `checkPurity` twice, though.